src/validation.cpp

```diff
@@ -4674,29 +4674,39 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                 break;
             }
             try {
-                // read block
+                // read block header
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
-                blkdat.SetLimit(nBlockPos + nSize);
-                blkdat.SetPos(nBlockPos);
-                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                CBlock& block = *pblock;
-                blkdat >> block;
-                nRewind = blkdat.GetPos();
-
-                uint256 hash = block.GetHash();
+                blkdat.SetLimit(nBlockPos + 80);
+                CBlockHeader header;
+                blkdat >> header;
+
+                const uint256 hash = header.GetHash();
                 {
                     LOCK(cs_main);
                     // detect out of order blocks, and store them for later
-                    if (hash != chainparams.GetConsensus().hashGenesisBlock && !LookupBlockIndex(block.hashPrevBlock)) {
+                    if (hash != chainparams.GetConsensus().hashGenesisBlock && !LookupBlockIndex(header.hashPrevBlock)) {
                         LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
-                                block.hashPrevBlock.ToString());
+                                header.hashPrevBlock.ToString());
                         if (dbp)
-                            mapBlocksUnknownParent.insert(std::make_pair(block.hashPrevBlock, *dbp));
+                            mapBlocksUnknownParent.insert(std::make_pair(header.hashPrevBlock, *dbp));
+
+                        // Position to the start of the next block.
+                        nRewind = nBlockPos + nSize;
+                        blkdat.SetLimit(nRewind);
+                        blkdat.Skip(nRewind - blkdat.GetPos());
```

Comment:
Skip() calls read(), which means (IIUC) that this change makes it so we are now doing a disk read while holding cs_main. Given this is reindexing this might not matter but worth noting. Is this a problem?

Reply:
Also, I think this part could be more concise. I mean adding position + size just do subtract position below, seems a bit redundant.

Reply:
> Skip() calls read()

Good catch, and it's actually even worse, because if we process the block immediately (don't take the `Skip()` path), then we deserialize the entire block (`blkdat >> block`, just below), and that calls `read()` which can read the disk. But that's the less common path (only about 8%).

As someone said at review club, the node isn't fully up and running during reindexing, so there probably isn't much contention for this lock. But I think we can avoid this problem; I'll push a fix if it doesn't complicate the code too much.

> this part could be more concise. I mean adding position + size just do subtract position below, seems a bit redundant.

I don't think it's redundant; note that `nBlockPos` is the start of the current block (whose header we just read), while `blkdat.GetPos()` returns the current position, which is _after_ the header. We don't want to skip `nSize` bytes (which is the length of the block), we want to skip the _remainder_ of the current block, so the distance to move forward is the start of the next block minus the current position. This could have been written `blkdat.Skip(nSize - 80)` but I think that's not as clear.

I just noticed that the `SetLimit()` here can be removed; it's not needed when doing a `Skip()` (it's only needed when deserializing something of variable length, such as a block). I'll remove it.